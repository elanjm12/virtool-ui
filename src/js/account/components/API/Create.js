import { mapValues } from "lodash-es";
import React, { useEffect, useState } from "react";
import { CopyToClipboard } from "react-copy-to-clipboard";
import { connect } from "react-redux";
import styled from "styled-components";
import { pushState } from "../../../app/actions";

import {
    ModalBody,
    ModalFooter,
    Icon,
    Input,
    InputContainer,
    InputError,
    InputGroup,
    InputIcon,
    InputLabel,
    Modal,
    SaveButton,
    ModalHeader
} from "../../../base";
import { routerLocationHasState } from "../../../utils/utils";
import { clearAPIKey, createAPIKey } from "../../actions";
import CreateAPIKeyInfo from "./CreateInfo";
import APIPermissions from "./Permissions";
import { Formik, Form, Field } from "formik";
import * as Yup from "yup";

const CreateAPIKeyCopied = styled.p`
    color: ${props => props.theme.color.blue};
`;

const CreateAPIKeyInput = styled(Input)`
    text-align: center;
`;

const StyledCreateAPIKey = styled(ModalBody)`
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding-left: 10px;
    padding-right: 10px;
    text-align: center;

    strong {
        color: ${props => props.theme.color.greenDark};
        margin-bottom: 5px;
    }
`;

const APIKeyInputContainer = styled(InputContainer)`
    margin-top: 15px;
    margin-bottom: 10px;
`;

const validationSchema = Yup.object().shape({
    name: Yup.string().required("Provide a name for the key")
});

export const getInitialFormValues = props => ({
    name: "",
    permissions: mapValues(props.permissions, () => false)
});

export const CreateAPIKey = props => {
    const [copied, setCopied] = useState(false);
    const [show, setShow] = useState(false);

    useEffect(() => {
        if (!show && props.newKey) {
            setShow(true);
        }
    }, [props.newKey]);

    const handleModalExited = () => {
        setCopied(false);
        setShow(false);
    };

    const handleSubmit = ({ name, permissions }) => {
        props.onCreate(name, permissions);
    };

    let content;

    if (show) {
        content = (
            <StyledCreateAPIKey>
                <strong>Here is your key.</strong>
                <small>Make note of it now. For security purposes, it will not be shown again.</small>

                <APIKeyInputContainer align="right">
                    <CreateAPIKeyInput value={props.newKey} readOnly />
                    <CopyToClipboard text={props.newKey} onCopy={() => setCopied(true)}>
                        <InputIcon aria-label="copy" name="copy" />
                    </CopyToClipboard>
                </APIKeyInputContainer>
                {copied && (
                    <CreateAPIKeyCopied>
                        <Icon name="check" /> Copied
                    </CreateAPIKeyCopied>
                )}
            </StyledCreateAPIKey>
        );
    } else {
        content = (
            <>
                <ModalHeader>Create API Key</ModalHeader>
                <Formik
                    onSubmit={handleSubmit}
                    initialValues={getInitialFormValues(props)}
                    validationSchema={validationSchema}
                >
                    {({ errors, setFieldValue, touched, values }) => (
                        <Form>
                            <CreateAPIKeyInfo />
                            <ModalBody>
                                <InputGroup>
                                    <InputLabel htmlFor="name">Name</InputLabel>
                                    <Field name="name" id="name" as={Input} />
                                    <InputError>{touched.name && errors.name}</InputError>
                                </InputGroup>

                                <label>Permissions</label>

                                <APIPermissions
                                    keyPermissions={values.permissions}
                                    onChange={(key, value) =>
                                        setFieldValue("permissions", { ...values.permissions, [key]: value })
                                    }
                                />
                            </ModalBody>

                            <ModalFooter>
                                <SaveButton />
                            </ModalFooter>
                        </Form>
                    )}
                </Formik>
            </>
        );
    }

    return (
        <Modal label="Create API Key" show={props.show} onHide={props.onHide} onExited={handleModalExited}>
            {content}
        </Modal>
    );
};

export const mapStateToProps = state => {
    return {
        show: routerLocationHasState(state, "createAPIKey"),
        newKey: state.account.newKey,
        permissions: state.account.permissions
    };
};

export const mapDispatchToProps = dispatch => ({
    onCreate: (name, permissions) => {
        dispatch(createAPIKey(name, permissions));
    },

    onHide: () => {
        dispatch(pushState({ createAPIKey: false }));
        dispatch(clearAPIKey());
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateAPIKey);
