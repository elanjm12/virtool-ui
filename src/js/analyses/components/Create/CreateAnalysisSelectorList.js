import React from "react";
import styled from "styled-components";
import { BoxGroup, BoxGroupSection } from "../../../base";
import { getBorder } from "../../../app/theme";

const StyledCreateAnalysisSelectorList = styled(BoxGroup)`
    border: none;
    background-color: ${props => props.theme.color.greyLightest};
    margin: 0;
    overflow-y: auto;
    height: 160px;

    ${BoxGroupSection} {
        border-bottom: ${getBorder};
    }
`;
export const CreateAnalysisSelectorList = ({ className, render, items }) => (
    <StyledCreateAnalysisSelectorList className={className}>
        {items.map(item => render(item))}
    </StyledCreateAnalysisSelectorList>
);
