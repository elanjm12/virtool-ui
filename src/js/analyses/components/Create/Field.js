import styled from "styled-components";
import { getFontSize } from "../../../app/theme";

export const CreateAnalysisField = styled.div`
    align-items: stretch;
    column-gap: ${props => props.theme.gap.column};
    display: grid;
    grid-template-columns: 1fr 1fr;
`;

export const CreateAnalysisFieldTitle = styled.div`
    font-size: ${getFontSize("lg")};
    grid-column: 1 / 3;
    margin-bottom: 10px;
`;
