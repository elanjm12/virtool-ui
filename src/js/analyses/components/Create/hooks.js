import { useEffect, useState } from "react";

export const useCreateAnalysis = (dataType, defaultSubtractions) => {
    const [errors, setErrors] = useState({});
    const [indexes, setIndexes] = useState([]);
    const [subtractions, setSubtractions] = useState(defaultSubtractions);
    const [workflows, setWorkflows] = useState([]);

    const setIndexesAndError = indexes => {
        setIndexes(indexes);
        setErrors({
            ...errors,
            indexes: false
        });
    };

    const setWorkflowsAndError = workflows => {
        setWorkflows(workflows);
        setErrors({
            ...errors,
            workflows: false
        });
    };

    useEffect(() => {
        setErrors({});
        setIndexes([]);
        setWorkflows([]);
    }, [dataType]);

    return {
        errors,
        indexes,
        subtractions,
        workflows,
        setErrors,
        setIndexes: setIndexesAndError,
        setSubtractions,
        setWorkflows: setWorkflowsAndError
    };
};
