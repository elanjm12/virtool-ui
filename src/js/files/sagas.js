import { noop } from "lodash-es";
import { buffers, END, eventChannel } from "redux-saga";
import { call, put, take, takeEvery, throttle } from "redux-saga/effects";
import { FIND_FILES, REMOVE_FILE, UPLOAD } from "../app/actionTypes";
import { apiCall, putGenericError } from "../utils/sagas";
import { uploadFailed, uploadProgress } from "./actions";
import * as filesAPI from "./api";

export function* watchFiles() {
    yield throttle(100, FIND_FILES.REQUESTED, findFiles);
    yield takeEvery(REMOVE_FILE.REQUESTED, removeFile);
    yield takeEvery(UPLOAD.REQUESTED, upload);
}

export function* findFiles(action) {
    yield apiCall(filesAPI.list, action.payload, FIND_FILES, {
        fileType: action.payload.fileType
    });
}

export function* removeFile(action) {
    yield apiCall(filesAPI.remove, action.payload, REMOVE_FILE);
}

export function* upload(action) {
    const { localId } = action.payload;
    const channel = yield call(createUploadChannel, action.payload, filesAPI.upload);
    yield watchUploadChannel(channel, UPLOAD, localId);
}

export const createUploadChannel = (action, apiMethod) =>
    eventChannel(emitter => {
        const onProgress = e => {
            if (e.lengthComputable) {
                const { total, loaded, percent } = e;
                emitter({ progress: percent, total, loaded });
            }
        };

        const onSuccess = response => {
            emitter({ response });
            emitter(END);
        };

        const onFailure = err => {
            emitter({ err });
            emitter(END);
        };

        apiMethod({
            ...action,
            onProgress,
            onSuccess,
            onFailure
        });

        return noop;
    }, buffers.sliding(2));

export function* watchUploadChannel(channel, actionType, localId) {
    let lastLoaded = 0;

    const loadedWindow = [];
    let count = 0;
    let uploadSpeed = 0;

    const intervalId = setInterval(() => {
        let loadedDuringInterval = 0;
        if (count < 6) {
            loadedWindow.push(lastLoaded);
            loadedDuringInterval = loadedWindow[loadedWindow.length - 1];
            count++;
        } else {
            loadedWindow.shift();
            loadedWindow.push(lastLoaded);
            loadedDuringInterval = loadedWindow[loadedWindow.length - 1] - loadedWindow[0];
        }
        uploadSpeed = loadedDuringInterval / count;
    }, 1000);

    while (true) {
        const { progress = 0, total, loaded, response, err } = yield take(channel);

        lastLoaded = loaded;

        if (err) {
            yield put(uploadFailed(localId));
            yield putGenericError(actionType, err);
            break;
        }
        if (response) {
            yield put({
                type: actionType.SUCCEEDED,
                payload: {
                    ...response.body,
                    localId
                }
            });
            break;
        }

        let remaining = (total - loaded) / uploadSpeed;
        remaining = isFinite(remaining) ? remaining : 0;

        yield put(uploadProgress(localId, progress, uploadSpeed, remaining));
    }

    clearInterval(intervalId);
}
