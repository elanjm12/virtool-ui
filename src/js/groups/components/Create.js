import React from "react";
import { pushState } from "../../app/actions";
import { getRouterLocationStateValue } from "../../app/selectors";
import {
    Modal,
    ModalBody,
    ModalFooter,
    ModalHeader,
    InputGroup,
    InputLabel,
    InputError,
    Input,
    SaveButton
} from "../../base";
import { Formik, Form, Field } from "formik";
import * as Yup from "yup";
import { connect } from "react-redux";
import { createGroup } from "../actions";
import { get } from "lodash-es";

const validationSchema = Yup.object().shape({
    name: Yup.string().required("Provide a name for the group")
});

export const Create = ({ onCreate, show, onHide, error }) => {
    const handleSubmit = values => {
        onCreate(values.name);
    };
    return (
        <Modal label="Create Group" onHide={onHide} show={show} size="lg">
            <ModalHeader>Create Group</ModalHeader>
            <Formik onSubmit={handleSubmit} initialValues={{ name: "" }} validationSchema={validationSchema}>
                {({ errors, touched }) => (
                    <Form>
                        <ModalBody>
                            <InputGroup>
                                <InputLabel>Name</InputLabel>
                                <Field name="name" id="name" as={Input} />
                                <InputError>{touched.name && (error || errors.name)}</InputError>
                            </InputGroup>
                        </ModalBody>
                        <ModalFooter>
                            <SaveButton />
                        </ModalFooter>
                    </Form>
                )}
            </Formik>
        </Modal>
    );
};

const mapStateToProps = state => ({
    show: Boolean(getRouterLocationStateValue(state, "createGroup")),
    error: get(state, "errors.CREATE_GROUP_ERROR.message", "")
});

const mapDispatchToProps = dispatch => ({
    onCreate: groupId => {
        dispatch(createGroup(groupId));
    },
    onHide: () => {
        dispatch(pushState({ createGroup: false }));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(Create);
