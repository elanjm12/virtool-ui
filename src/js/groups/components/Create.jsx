import React from "react";
import { pushState } from "../../app/actions";
import { getRouterLocationStateValue } from "../../app/selectors";
import { Field, Form, Formik } from "formik";
import * as Yup from "yup";
import { connect } from "react-redux";
import { createGroup } from "../actions";
import { get } from "lodash-es";
import {
    Input,
    InputError,
    InputGroup,
    InputLabel,
    Modal,
    ModalBody,
    ModalFooter,
    ModalHeader,
    SaveButton
} from "../../base";

const validationSchema = Yup.object().shape({
    name: Yup.string().required("Provide a name for the group")
});

export const Create = ({ error, show, onCreate, onHide }) => {
    const handleSubmit = values => {
        onCreate(values.name);
    };
    return (
        <Modal label="Create" onHide={onHide} show={show} size="sm">
            <ModalHeader>Create Group</ModalHeader>
            <Formik onSubmit={handleSubmit} initialValues={{ name: "" }} validationSchema={validationSchema}>
                {({ errors, touched }) => (
                    <Form>
                        <ModalBody>
                            <InputGroup>
                                <InputLabel>Name</InputLabel>
                                <Field name="name" id="name" as={Input} />
                                <InputError>{touched.name && (error || errors.name)}</InputError>
                            </InputGroup>
                        </ModalBody>
                        <ModalFooter>
                            <SaveButton />
                        </ModalFooter>
                    </Form>
                )}
            </Formik>
        </Modal>
    );
};

const mapStateToProps = state => ({
    show: Boolean(getRouterLocationStateValue(state, "createGroup")),
    error: get(state, "errors.CREATE_GROUP_ERROR.message", "")
});

const mapDispatchToProps = dispatch => ({
    onCreate: groupId => {
        dispatch(createGroup(groupId));
    },
    onHide: () => {
        dispatch(pushState({ createGroup: false }));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(Create);
