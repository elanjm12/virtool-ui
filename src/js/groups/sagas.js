import { push } from "connected-react-router";
import { some } from "lodash-es";
import { takeEvery, takeLatest, throttle, select, put } from "redux-saga/effects";
import {
    CREATE_GROUP,
    GET_GROUP,
    LIST_GROUPS,
    REMOVE_GROUP,
    SET_GROUP_PERMISSION,
    WS_REMOVE_GROUP
} from "../app/actionTypes";
import { apiCall } from "../utils/sagas";
import * as groupsAPI from "./api";
import { getActiveGroup, getGroups } from "./selectors";

export function* watchGroups() {
    yield takeLatest(LIST_GROUPS.REQUESTED, listGroups);
    yield throttle(200, CREATE_GROUP.REQUESTED, createGroup);
    yield takeEvery(SET_GROUP_PERMISSION.REQUESTED, setGroupPermission);
    yield throttle(100, REMOVE_GROUP.REQUESTED, removeGroup);
    yield takeLatest(GET_GROUP.REQUESTED, getGroup);
    yield takeLatest(WS_REMOVE_GROUP, UpdateActiveGroup);
}

function* listGroups(action) {
    const resp = yield apiCall(groupsAPI.list, action, LIST_GROUPS);
    if (resp.ok) {
        yield UpdateActiveGroup();
    }
}

function* UpdateActiveGroup() {
    const activeGroup = yield select(getActiveGroup);
    const groups = yield select(getGroups);
    if (!activeGroup || !some(groups, { id: activeGroup.id })) {
        if (groups.length) yield getGroup({ payload: { groupId: groups[0].id } });
    }
}

function* getGroup(action) {
    yield apiCall(groupsAPI.get, action.payload, GET_GROUP);
}

function* createGroup(action) {
    const resp = yield apiCall(groupsAPI.create, action.payload, CREATE_GROUP);

    if (resp.ok) {
        yield put(push({ state: { createGroup: false } }));
    }
}

function* setGroupPermission(action) {
    yield apiCall(groupsAPI.setPermission, action.payload, SET_GROUP_PERMISSION);
}

function* removeGroup(action) {
    yield apiCall(groupsAPI.remove, action.payload, REMOVE_GROUP);
}
