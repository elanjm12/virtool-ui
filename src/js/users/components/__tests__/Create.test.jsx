import { PUSH_STATE } from "../../../app/actionTypes";
import { CreateUser, mapDispatchToProps, mapStateToProps } from "../Create";
import { fireEvent, screen } from "@testing-library/react";
import { ConnectedRouter, connectRouter, routerMiddleware } from "connected-react-router";
import { createBrowserHistory } from "history";
import React from "react";
import { Provider } from "react-redux";
import createSagaMiddleware from "redux-saga";
import { applyMiddleware, combineReducers, createStore } from "redux";
import { watchRouter } from "../../../app/sagas";

const createAppStore = (state, history) => {
    const reducer = combineReducers({
        router: connectRouter(history)
    });
    const sagaMiddleware = createSagaMiddleware();
    const store = createStore(reducer, applyMiddleware(sagaMiddleware, routerMiddleware(history)));

    sagaMiddleware.run(watchRouter);

    return store;
};

const renderWithRouter = (ui, state, history) => {
    const wrappedUI = (
        <Provider store={createAppStore(state, history)}>
            <ConnectedRouter history={history}> {ui} </ConnectedRouter>
        </Provider>
    );
    renderWithProviders(wrappedUI);
};

describe("<CreateUser />", () => {
    let props;
    let state;

    beforeEach(() => {
        props = {
            error: "",
            show: true,
            minimumPasswordLength: 8,
            onClearError: vi.fn(),
            onCreate: vi.fn(),
            onHide: vi.fn()
        };

        state = {
            errorPassword: "",
            errorHandle: "",
            forceReset: false,
            password: "",
            handle: ""
        };
        history = createBrowserHistory();
    });

    it("should render correctly when show=true", () => {
        renderWithRouter(<CreateUser {...props} />, state, history);
        expect(screen.getByText("Create User")).toBeInTheDocument();
        expect(screen.getByText("Username")).toBeInTheDocument();
        expect(screen.getByText("Password")).toBeInTheDocument();
        expect(screen.getByText("Force user to reset password on login"));
        expect(screen.getByRole("button", { name: "Save" })).toBeInTheDocument();
    });

    it("should render correct error messages", () => {
        renderWithRouter(<CreateUser {...props} />, state, history);
        const saveButton = screen.getByRole("button", { name: "Save" });
        saveButton.click();
        expect(screen.getByText("Please specify a username")).toBeInTheDocument();
        expect(screen.getByText("Passwords must contain at least 8 characters")).toBeInTheDocument();
        expect(props.onCreate).not.toHaveBeenCalled();
    });

    it("should not display error messages and should call onCreate when valid info entered", () => {
        renderWithRouter(<CreateUser {...props} />, state, history);
        const userNameInput = screen.getByLabelText("username");
        const passwordInput = screen.getByLabelText("password");
        const saveButton = screen.getByRole("button", { name: "Save" });
        fireEvent.change(userNameInput, { target: { value: "newUsername" } });
        fireEvent.change(passwordInput, { target: { value: "newPassword" } });
        expect(props.onCreate).not.toHaveBeenCalled();
        saveButton.click();
        expect(props.onCreate).toHaveBeenCalled();
    });

    it("should call onHide when model exited", () => {
        props.error = "Error";
        renderWithRouter(<CreateUser {...props} />, state, history);
        const closeButton = screen.getByLabelText("close");
        expect(props.onHide).toHaveBeenCalledTimes(0);
        fireEvent.click(closeButton);
        expect(props.onHide).toHaveBeenCalledTimes(1);
    });
});

describe("mapStateToProps", () => {
    it("should call mapStateToProps", () => {
        const state = {
            errors: {
                CREATE_USER_ERROR: {
                    message: "foo"
                }
            },
            router: { location: "foo" },
            users: {
                createPending: true
            },
            settings: {
                data: {
                    minimum_password_length: 1
                }
            }
        };

        const result = mapStateToProps(state);
        expect(result).toEqual({
            show: false,
            pending: true,
            minimumPasswordLength: 1,
            error: "foo"
        });
    });
});

describe("mapDispatchToProps", () => {
    let dispatch;
    let result;

    beforeEach(() => {
        dispatch = vi.fn();
        result = mapDispatchToProps(dispatch);
    });

    it("should return onCreate() in props", () => {
        const data = "foo";

        result.onCreate(data);
        expect(dispatch).toHaveBeenCalledWith({
            payload: "foo",
            type: "CREATE_USER_REQUESTED"
        });
    });
    it("should return onHide() in props", () => {
        result.onHide();
        expect(dispatch).toHaveBeenCalledWith({ type: PUSH_STATE, payload: { state: { createUser: false } } });
    });

    it("should return onClearError() in props", () => {
        result.onClearError();
        expect(dispatch).toHaveBeenCalledWith({
            payload: { error: "CREATE_USER_ERROR" },
            type: "CLEAR_ERROR"
        });
    });
});
