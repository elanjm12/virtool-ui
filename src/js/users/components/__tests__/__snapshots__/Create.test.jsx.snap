// Vitest Snapshot v1

exports[`<CreateUser /> > should render 1`] = `ShallowWrapper {}`;

exports[`<CreateUser /> > should render when name has changed 1`] = `ShallowWrapper {}`;

exports[`<CreateUser /> > should render when name has changed 2`] = `ShallowWrapper {}`;

exports[`<CreateUser /> > should render when password has changed 1`] = `ShallowWrapper {}`;

exports[`<CreateUser /> > should render when password has changed 2`] = `ShallowWrapper {}`;

exports[`<CreateUser /> > should render with error 1`] = `ShallowWrapper {}`;
